src/graph/network/msg_queue.cc

```diff
@@ -0,0 +1,190 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file msg_queue.cc
+ * \brief Message queue for DGL distributed training.
+ */
+#include <dmlc/logging.h>
+#include <cstring>
+
+#include "msg_queue.h"
+
+namespace dgl {
+namespace network {
+
+using std::string;
+
+MessageQueue::MessageQueue(int64_t queue_size, int num_producers) {
+  CHECK_LT(0, queue_size);
+  try {
+    queue_ = new char[queue_size];
+  } catch(const std::bad_alloc&) {
+    LOG(FATAL) << "Not enough memory for message queue.";
+  }
+  memset(queue_, '\0', queue_size);
+
+  queue_size_ = queue_size;
+  free_size_ = queue_size;
+  write_pointer_ = 0;
+  num_producers_ = num_producers;
+}
+
+MessageQueue::~MessageQueue() {
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (nullptr != queue_) {
+    delete [] queue_;
+    queue_ = nullptr;
+  }
+}
+
+int64_t MessageQueue::Add(const char* src, int64_t size, bool is_blocking) {
```

Comment:
I actually don't understand why this message queue implementation is preferred. Why not always allocate a new piece of memory to store received data and add the piece of memory into the message queue? In this way, there isn't two extra memory copy (one memory copy when adding to the queue, another memory copy when removing from the queue). Since NodeFlow can be quite large, the extra memory copy can cause significant overhead.

Reply:
Allocate new memory has the following disadvantages:

1. Allocate a new piece of memory is much more expensive than copy a piece of memory.
2. We should control the buffer size to avoid allocating too much memory when we have lot of remote samplers.